# Big Data Modeling and Management Assigment - Homework 1

# Submission

GROUP NUMBER: **12** - please add your group number into the file name

GROUP MEMBERS:

|STUDENT NAME|STUDENT NUMBER|
|---|---|
|Catarina Gonçalves Nunes|20230083|
|XXXXXX|XXXXXX|
|XXXXXX|XXXXXX|
|XXXXXX|XXXXXX|

## 🍺 The Beer project  🍺 

As it was shown in classes, graph databases are a natural way of navegating related information. For this first project we will be taking a graph database to analyse beer and breweries!   

The project datasets are based on [kaggle](https://www.kaggle.com/ehallmar/beers-breweries-and-beer-reviews), released by Evan Hallmark. 

### Problem description

Imagine you are working in the Data Management department of Analytics company.
Explore the database via python neo4j connector and/or the graphical tool in the NEO4J webpage. Answer the questions while adjusting the database to meet the needs of your colleagues.
Please record and keep track of your database changes, and submit the file with all cells run and with the output shown.

### Questions

1. Explore the database: get familiar with current schema, elements and other important database parameters. [1 point]
2. Adjust the database and mention reasoning behind: e.g. clean errors, remove redundancies, adjust schema as necessary. Visualize the final version of database schema. [4 points]
3. Analytics department requires the following information for the biweekly reporting: [5 points]
    1. How many reviews has the beer with the most reviews?
    2. Which three users wrote the most reviews about beers?
    3. Find all beers that are described with following words: 'fruit', 'complex', 'nutty', 'dark'.
    4. Which top three breweries produce the largest variety of beer styles?
    5. Which country produces the most beer styles?
4. Market Analysis department in your company accesses and updates the trends data on the daily basis. Given that, consider how you need to optimize the database and its performance so that the following queries are efficient. Measure performance to communicate your improvements using PROFILE before final query. Answer the following: [4 points]
    1. Using ABV score, find five strongest beers, display their ABV score and the corresponding brewery? Keep in mind that the strongest known beer is Snake Venom, and deal with the error entries in the database.
    2. Using the answer from question 2, find the top 5 distict beer styles with the highest average score of smell + feel that were reviewed by the third most productive user. Keep in mind that cleaning the database earlier should ensure correct results.
5. Answer **two out of four** of the following questions using Graph Algorithms (gds): [NB: make sure to clear the graph before using it again] For the quarterly report, Analytics department the follownig information. [6 points]
    1. Which two countries are most similiar when it comes to their top five most produced Beer styles?
    2. Which beer is the most popular when considering the number of users who reviewed it? 
    3. Users are connected together by their reviews of beers, taking into consideration the "smell" score they assign as a weight, how many communities are formed from these relationships? How many users are in the three largest communities? 
    4. Which user is the most influential when it comes to reviews of distinct beers by style?
 
### Groups  

Groups should have 4 people maximum. Please mark which group you are here: https://shorturl.at/zE0QP 

### Submission      

The code used to produce the results and to-the-point explations should be uploaded to moodle. They should have a clear reference to the group, either on the file name or on the document itself. Preferably one Jupyter notebook per group.

Delivery date: Until the **midnight of March 18, 2025**

### Evaluation   

This will be 20% of the final grade.   
Each solution will be evaluated on 2 components: correctness of results and efficiency of the query (based on database schema).  
All code will go through plagiarism automated checks. Groups with the same code will undergo investigation.

## Loading the Database

#### Be sure that you **don't have** the neo4j docker container from the classes running (you can Stop it in the desktop app or with the command "`docker stop Neo4JLab`")


The default container does not have any data whatsoever, we will have to load a database into our docker image:
- Download and unzip the `Neo4JHWData` file provided in Moodle.
- Copy the path of the `Neo4JHWData` folder of the unziped file, e.g. `C:/PATH/Neo4JHWData/data`.
- Download and unzip the `Neo4JPlugins` file provided in Moodle.
- Copy the path of the `Neo4JPlugins` folder of the unziped file, e.g. `C:/PATH/Neo4Jplugins`.
- Change the code below accordingly. As you might have noticed, you do not have a user called `nunoa`, please use the appropriate path that you got from the previous step. Be sure that you have a neo4j docker container running: \

`docker run --name Neo4JHW2025 -p 7474:7474 -p 7687:7687 -d -v "c:\PATH\Neo4JPlugins":/plugins -v "c:\PATH\Neo4JHWData\data":/data --env NEO4J_AUTH=neo4j/test --env NEO4J_dbms_connector_https_advertised__address="localhost:7473" --env NEO4J_dbms_connector_http_advertised__address="localhost:7474" --env NEO4J_dbms_connector_bolt_advertised__address="localhost:7687" --env NEO4J_dbms_security_procedures_unrestricted=gds.* --env NEO4J_dbms_security_procedures_allowlist="gds.*" neo4j:4.4.5`

- Since Neo4j is trying to recognize a new database folder, this might take a bit (let's say 3 minutes), so don't worry.

If the neo4j browser fails to load gds plugins, run the following in the Command Prompt before creating the container again:
`// Remove stopped containers //
docker container prune -f
// Remove unused images //
docker image prune -a -f
// Remove unused volumes //
docker volume prune -f
// Remove unused networks //
docker network prune -f
// Remove all unused resources in one command //
docker system prune -a -f`

## Index

1. [Explore the Database: Understanding the Current Schema, Elements, and Important Parameters](#explore-the-database)
2. [Adjust the Database and Mention the Reasoning Behind the Adjustments](#adjust-the-database) 
3. [Analytics Department Biweekly Reporting Requirements](#analytics-department)
   1. [How many reviews does the beer with the most reviews have?](#beer-with-most-reviews)
   2. [Which three users wrote the most reviews about beers?](#three-users-most-reviews)
   3. [Find all beers that are described with the following words: 'fruit', 'complex', 'nutty', 'dark'.](#beers-with-descriptions)
   4. [Which top three breweries produce the largest variety of beer styles?](#top-three-breweries)
   5. [Which country produces the most beer styles?](#country-most-styles)
4. [Optimizing Database Performance for Market Analysis Department](#optimizing-performance)
   1. [Using ABV score, find five strongest beers, display their ABV score and the corresponding brewery.](#strongest-beers)
   2. [Find the top 5 distinct beer styles with the highest average score of smell and feel that were reviewed by the third most productive user.](#top-5-beer-styles)
5. [Answer Two Out of Four Questions Using Graph Algorithms (gds)](##graph-algorithms)
   1. [Which two countries are most similar when it comes to their top five most produced Beer styles?](#similar-countries)
   2. [Which beer is the most popular considering the number of users who reviewed it?](#most-popular-beer)
   3. [How many communities are formed from these relationships? How many users are in the three largest communities?](#user-communities)
   4. [Which user is the most influential when it comes to reviews of distinct beers by style?](#influential-user)

### <div class="alert-warning"> <a id="explore-the-database"></a>1. Explore the Database: Understanding the Current Schema, Elements, and Important Parameters </div>

In [3]:
#! pip install neo4j

In [1]:
from neo4j import GraphDatabase
from pprint import pprint

In [2]:
NEO4J_URI="neo4j://localhost:7687"
NEO4J_USERNAME="neo4j"
NEO4J_PASSWORD="test"

In [3]:
driver = GraphDatabase.driver(NEO4J_URI, auth=(NEO4J_USERNAME, NEO4J_PASSWORD), )

In [4]:
def execute_read(driver, query):    
    with driver.session(database="neo4j") as session:
        result = session.execute_read(lambda tx, query: list(tx.run(query)), query)
    return result

1.1 The dataset consists of the following nodes (labels):

In [5]:
query = """
       call db.labels();
    """

result = execute_read(driver, query)

pprint(result)

[<Record label='COUNTRIES'>,
 <Record label='CITIES'>,
 <Record label='BREWERIES'>,
 <Record label='BEERS'>,
 <Record label='REVIEWS'>,
 <Record label='STYLE'>,
 <Record label='USER'>]


1.2. Relationships between nodes:

In [6]:
query = """
        CALL db.relationshipTypes();
    """

result = execute_read(driver, query)

pprint(result)

[<Record relationshipType='REVIEWED'>,
 <Record relationshipType='BREWED'>,
 <Record relationshipType='IN'>,
 <Record relationshipType='HAS_STYLE'>,
 <Record relationshipType='POSTED'>]


In [7]:
query = """
    MATCH (n)
    RETURN labels(n) AS tipo, COUNT(n) AS total
    ORDER BY total DESC;
"""
result = execute_read(driver, query)
pprint(result)

[<Record tipo=['REVIEWS'] total=2549252>,
 <Record tipo=['BEERS'] total=417746>,
 <Record tipo=['USER'] total=123935>,
 <Record tipo=['BREWERIES'] total=100694>,
 <Record tipo=['CITIES'] total=23330>,
 <Record tipo=['COUNTRIES'] total=400>,
 <Record tipo=['STYLE'] total=113>]


> REVIEWS is the most common node (makes sense, since a user can make reviews about multiple beers).
USER count is ~124K, meaning on average, each user has written ~20 reviews (2549252 / 123935).
BREWERIES count is ~100K, which seems high—there might be duplicate or incorrect data.
CITIES (~23K) and COUNTRIES (400) seem reasonable.
STYLE has only 113, which makes sense (beer styles are limited, e.g., IPA, Stout, Lager).

In [17]:
query = """
    CALL db.schema.visualization();
"""
result = execute_read(driver, query)
pprint(result)

[<Record nodes=[<Node element_id='-5' labels=frozenset({'REVIEWS'}) properties={'name': 'REVIEWS', 'indexes': ['id'], 'constraints': []}>, <Node element_id='-4' labels=frozenset({'BEERS'}) properties={'name': 'BEERS', 'indexes': ['id'], 'constraints': []}>, <Node element_id='-1' labels=frozenset({'COUNTRIES'}) properties={'name': 'COUNTRIES', 'indexes': ['name'], 'constraints': []}>, <Node element_id='-3' labels=frozenset({'BREWERIES'}) properties={'name': 'BREWERIES', 'indexes': ['id'], 'constraints': []}>, <Node element_id='-6' labels=frozenset({'STYLE'}) properties={'name': 'STYLE', 'indexes': ['name'], 'constraints': []}>, <Node element_id='-2' labels=frozenset({'CITIES'}) properties={'name': 'CITIES', 'indexes': ['name'], 'constraints': []}>, <Node element_id='-7' labels=frozenset({'USER'}) properties={'name': 'USER', 'indexes': ['name'], 'constraints': []}>] relationships=[<Relationship element_id='-1' nodes=(<Node element_id='-4' labels=frozenset({'BEERS'}) properties={'name': '

From the schema visualization, the database contains the following nodes and relationships:

CITIES → Connected to BREWERIES (relationship: IN)
BREWERIES → Connected to BEERS (relationship: BREWED)
BEERS → Connected to STYLE (HAS_STYLE) and REVIEWS (REVIEWED)
REVIEWS → Connected to USER (POSTED)- 

![Database Schema Visualization](https://raw.githubusercontent.com/CatarinaGN/Big-Data-Modelling-and-Management/main/db.schema.visualization().png)


In [8]:
query = """
    MATCH (b:BEERS)
    RETURN b.name, b.ABV, b.rating LIMIT 10;
"""
result = execute_read(driver, query)
pprint(result)

[<Record b.name='Olde Cogitator' b.ABV=None b.rating=None>,
 <Record b.name='Konrads Stout Russian Imperial Stout' b.ABV=None b.rating=None>,
 <Record b.name='Scottish Right' b.ABV=None b.rating=None>,
 <Record b.name='MegaMeow Imperial Stout' b.ABV=None b.rating=None>,
 <Record b.name='Peaches-N-Cream' b.ABV=None b.rating=None>,
 <Record b.name='World Burp Beer 2002' b.ABV=None b.rating=None>,
 <Record b.name='Icon Sender' b.ABV=None b.rating=None>,
 <Record b.name='Divina IPA' b.ABV=None b.rating=None>,
 <Record b.name='Light Of The Ozarks' b.ABV=None b.rating=None>,
 <Record b.name="Warrior's Bock" b.ABV=None b.rating=None>]


In [19]:
query = """
    MATCH (u:USER)-[r:REVIEWED]->(b:BEERS)
    RETURN u.name AS user, b.name AS beer, r.rating, r.review_date LIMIT 10;
"""
result = execute_read(driver, query)
pprint(result)


[]


### <div class="alert-warning"> <a id="adjust-the-database"></a>2. Adjust the Database and Mention the Reasoning Behind the Adjustments </div>

### <a id="analytics-department"></a>3. Analytics Department Biweekly Reporting Requirements

#### <a id="beer-with-most-reviews"></a>3.1. How many reviews does the beer with the most reviews have?

#### <a id="three-users-most-reviews"></a>3.2. Which three users wrote the most reviews about beers?

#### <a id="beers-with-descriptions"></a>3.3. Find all beers that are described with the following words: 'fruit', 'complex', 'nutty', 'dark'.

#### <a id="top-three-breweries"></a>3.4. Which top three breweries produce the largest variety of beer styles?

#### <a id="country-most-styles"></a>3.5. Which country produces the most beer styles?